processor/deltatocumulativeprocessor/processor.go

```diff
@@ -32,66 +33,84 @@ type Processor struct {
 
 	log    *zap.Logger
 	ctx    context.Context
-	cancel context.CancelFunc
+	cancel context.CancelCauseFunc
 
-	aggr  streams.Aggregator[data.Number]
-	stale maybe.Ptr[staleness.Staleness[data.Number]]
+	sums Pipeline[data.Number]
+	expo Pipeline[data.ExpHistogram]
 
 	mtx sync.Mutex
 }
 
 func newProcessor(cfg *Config, log *zap.Logger, meter metric.Meter, next consumer.Metrics) *Processor {
-	ctx, cancel := context.WithCancel(context.Background())
+	ctx, cancel := fatal.Context(context.Background())
+
+	tel := telemetry.New(meter)
 
 	proc := Processor{
 		log:    log,
 		ctx:    ctx,
 		cancel: cancel,
 		next:   next,
+
+		sums: pipeline[data.Number](cfg, &tel),
+		expo: pipeline[data.ExpHistogram](cfg, &tel),
 	}
 
-	tel := telemetry.New(meter)
+	return &proc
+}
+
+type Pipeline[D data.Point[D]] struct {
+	aggr  streams.Aggregator[D]
+	stale maybe.Ptr[staleness.Staleness[D]]
+}
+
+func pipeline[D data.Point[D]](cfg *Config, tel *telemetry.Telemetry) Pipeline[D] {
+	var pipe Pipeline[D]
 
-	var dps streams.Map[data.Number]
-	dps = delta.New[data.Number]()
+	var dps streams.Map[D]
+	dps = delta.New[D]()
 	dps = telemetry.ObserveItems(dps, &tel.Metrics)
 
 	if cfg.MaxStale > 0 {
-		tel.WithStale(meter, cfg.MaxStale)
+		tel.WithStale(cfg.MaxStale)
 		stale := maybe.Some(staleness.NewStaleness(cfg.MaxStale, dps))
-		proc.stale = stale
+		pipe.stale = stale
 		dps, _ = stale.Try()
 	}
 	if cfg.MaxStreams > 0 {
-		tel.WithLimit(meter, int64(cfg.MaxStreams))
+		tel.WithLimit(int64(cfg.MaxStreams))
 		lim := streams.Limit(dps, cfg.MaxStreams)
-		if stale, ok := proc.stale.Try(); ok {
+		if stale, ok := pipe.stale.Try(); ok {
 			lim.Evictor = stale
 		}
 		dps = lim
 	}
 
 	dps = telemetry.ObserveNonFatal(dps, &tel.Metrics)
 
-	proc.aggr = streams.IntoAggregator(dps)
-	return &proc
+	pipe.aggr = streams.IntoAggregator(dps)
+	return pipe
 }
 
 func (p *Processor) Start(_ context.Context, _ component.Host) error {
-	stale, ok := p.stale.Try()
-	if !ok {
+	sums, sok := p.sums.stale.Try()
+	expo, eok := p.expo.stale.Try()
+	if !(sok && eok) {
 		return nil
 	}
 
 	go func() {
+		fatal.Recover(p.ctx)
```

Comment:
can we have a metric to track how many recovers we had? this should always be 0, but if it's not, we know there's something wrong here